pkg/networkserver/grpc_gsns.go

```diff
@@ -209,185 +204,186 @@ func (ns *NetworkServer) matchAndHandleDataUplink(ctx context.Context, dev *ttnp
-		if dev.PendingMACState.PendingJoinRequest.DownlinkSettings.OptNeg && dev.LoRaWANVersion.Compare(ttnpb.MAC_V1_1) >= 0 {
-			// The version will be further negotiated via RekeyInd/RekeyConf
-			dev.PendingMACState.LoRaWANVersion = ttnpb.MAC_V1_1
-		}
-		chs, ok := applyCFList(dev.PendingMACState.PendingJoinRequest.CFList, phy, dev.PendingMACState.CurrentParameters.Channels...)
-		if !ok {
-			log.FromContext(ctx).Debug("Failed to apply CFList, skip")
-			return nil, false, nil
-		}
-		dev.PendingMACState.CurrentParameters.Channels = chs
+				matchType = currentResetMatch
 
-		if dev.MACState.GetPendingApplicationDownlink() != nil {
-			queuedApplicationUplinks = []*ttnpb.ApplicationUp{
-				{
-					EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
-					Up: &ttnpb.ApplicationUp_DownlinkNack{
-						DownlinkNack: dev.MACState.PendingApplicationDownlink,
-					},
-					CorrelationIDs: append(dev.MACState.PendingApplicationDownlink.CorrelationIDs, up.CorrelationIDs...),
-				},
-			}
-			dev.MACState.PendingApplicationDownlink = nil
-		}
-		dev.MACState = dev.PendingMACState
-		dev.PendingSession.StartedAt = up.ReceivedAt
+			case cmacFMatchResult.FullFCnt > dev.Session.LastFCntUp:
+				ctx = log.NewContextWithField(ctx, "f_cnt_reset", false)
 
-		fCntGap = cmacFMatchResult.FullFCnt
-		session = dev.PendingSession
+				fCntGap := cmacFMatchResult.FullFCnt - dev.Session.LastFCntUp
+				if dev.MACState.LoRaWANVersion.HasMaxFCntGap() && uint(fCntGap) > phy.MaxFCntGap {
+					log.FromContext(ctx).WithFields(log.Fields(
+						"f_cnt_gap", fCntGap,
+						"max_f_cnt_gap", phy.MaxFCntGap,
+					)).Debug("FCnt gap exceeds maximum after reset")
+					return nil, false, nil
+				}
 
-	case fCntResetMatch:
-		ctx = log.NewContextWithField(ctx, "last_f_cnt_up", dev.Session.LastFCntUp)
-		if pld.Ack {
-			log.FromContext(ctx).Debug("ACK bit set in uplink after FCnt reset, skip")
-			return nil, false, nil
-		}
-		var err error
-		macState, err := mac.NewState(dev, ns.FrequencyPlans, ns.defaultMACSettings)
-		if err != nil {
-			log.FromContext(ctx).WithError(err).Warn("Failed to generate new MAC state")
-			return nil, false, nil
-		}
+				matchType = currentOriginalMatch
 
-		if dev.MACState.GetPendingApplicationDownlink() != nil {
-			queuedApplicationUplinks = []*ttnpb.ApplicationUp{
-				{
-					EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
-					Up: &ttnpb.ApplicationUp_DownlinkNack{
-						DownlinkNack: dev.MACState.PendingApplicationDownlink,
-					},
-					CorrelationIDs: append(dev.MACState.PendingApplicationDownlink.CorrelationIDs, up.CorrelationIDs...),
-				},
+			case dev.Session.LastFCntUp == 0 && dev.SupportsJoin && len(dev.MACState.RecentUplinks) == 1,
+				dev.Session.LastFCntUp == 0 && !dev.SupportsJoin && len(dev.MACState.RecentUplinks) == 0:
+
+				ctx = log.NewContextWithField(ctx, "f_cnt_reset", false)
+				matchType = currentOriginalMatch
+
+			default: // cmacFMatchResult.FullFCnt == dev.Session.LastFCntUp
```

Comment:
There are already 2 cases above:
`			case cmacFMatchResult.FullFCnt < dev.Session.LastFCntUp:`
`			case cmacFMatchResult.FullFCnt > dev.Session.LastFCntUp:`
Not sure what you mean and how I can improve this.
Can you give me a suggestion for a comment?
In this case FCnt is repeated and it's not the first uplink in the session

Reply:
You should also notice the `matchType`, which should clarify what's going on

Reply:
Mostly code comments would do

Reply:
Sorry, it's not clear to me. What do you mean exactly?
I'd rather not pollute code with redundant comments - if the code is not self-explanatory, then I'd rather fix the code. What is not clear in the implementation exactly?

This is a switch of 3 cases:
- A > B
- A < B
- A == B

I commented the `A==B` part, since that is hidden behind a `default:` statement.
What else should be explained?

Reply:
With the comment on this line, I was asking whether this code comment was a left-over or that it was descriptive.

With adding comments I didn't mean these cases [but these](https://github.com/TheThingsNetwork/lorawan-stack/pull/3603#discussion_r547894219). I see that you [already put comments there](https://github.com/rvolosatovs/lorawan-stack/blob/fix/matching/pkg/networkserver/grpc_gsns.go#L271-L276) so this thread can be resolved.